broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,939 @@
+    async fn handle_ack(
+        &mut self,
+        chunk: &wire::Chunk,
+        ack_epoch: Epoch,
+        partial: Bytes,
+    ) -> Result<(), Error> {
+        // If the ack is for an epoch that is too old or too new, ignore.
+        let (bound_lo, bound_hi) = {
+            let epoch = self.coordinator.index();
+            let (eb_lo, eb_hi) = self.epoch_bounds;
+            (epoch.saturating_sub(eb_lo), epoch.saturating_add(eb_hi))
+        };
+        if ack_epoch < bound_lo || ack_epoch > bound_hi {
+            return Err(Error::AckEpochOutsideBounds(ack_epoch, bound_lo, bound_hi));
+        }
+
+        // Get the number of required partials
+        let Some(identity) = self.coordinator.identity(ack_epoch) else {
+            return Err(Error::UnknownIdentity(ack_epoch));
+        };
+        let quorum = identity.required();
+
+        // Get the partial signatures, returning early if we already have a threshold
+        let evidence = self.ack_man.get_or_init(ack_epoch, chunk);
+        let partials = match evidence {
+            Evidence::Threshold(_) => return Ok(()),
+            Evidence::Partials(partials) => partials,
+        };
+
+        // Return early if we already have this partial
+        if partials.contains(&partial) {
+            debug!("Ignoring ack. Already have this partial.");
+            return Ok(());
+        }
+
+        // Store the ack
+        partials.insert(partial.clone());
+
+        // Return early if we don't have enough partials
+        if partials.len() < quorum as usize {
+            return Ok(());
+        }
+
+        // Construct the threshold signature
+        let partials: Vec<PartialSignature> = partials
+            .iter()
+            .map(|p| PartialSignature::deserialize(p).unwrap())
+            .collect();
+        let threshold = ops::threshold_signature_recover(quorum, partials).unwrap();
+
+        // Handle the threshold
+        self.handle_threshold(chunk, ack_epoch, threshold).await;
+
+        Ok(())
+    }
+
+    /// Handles a valid link message, storing it as the tip.
+    /// Alerts the application of the new link.
+    /// Also appends the link to the journal if it's new.
+    async fn handle_link(&mut self, link: &wire::Link) {
+        // Store the tip
+        let is_new = self.tip_man.put(link);
+        let chunk = link.chunk.as_ref().unwrap();
+
+        // Take actions if the link is new
+        if is_new {
+            // Append to journal
+            self.journal_append(link).await;
+            self.journal_sync(&chunk.sequencer, chunk.height).await;
+
+            // Prune old evidence
+            self.ack_man
+                .prune_height(self.epoch, &chunk.sequencer, chunk.height);
+        }
+
+        // Verify the chunk with the application
+        let context = Context {
+            sequencer: chunk.sequencer.clone(),
+            height: chunk.height,
+        };
+        self.application
```

Comment:
I would recommend doing a lot of the checks you perform after `verify` (in `handle_*`) before calling the application.

It should make us a lot more robust to incoming spam.